### ISSUE TITLE
Custom function descriptor in metaclass 

### ISSUE THREAD

**Issue by BrianPugh** [microsoft/pyright#3998]
**Describe the bug**
I would like to:

1.  Invoking `Class.method()` calls the **meta**class's bound method.
2. Invoking `Class().method()` calls the class's bound method.

The attached code snippet accomplishes this, but pyright returns: `Argument missing for parameter "self" (Pyright reportGeneralTypeIssues)` when invoking `Class.method()`.

**To Reproduce**
See code.

**Expected behavior**
I would expect the pyright error not to be raised. Well, really I understand that this would be really hard for a static checker to resolve, so I'm looking for help to implement this feature while still being pyright-friendly.

**Screenshots or Code**
```python
from types import MethodType


class MethodDescriptor(object):
    def __init__(self, method, meta_method):
        self.method = method
        self.meta_method = meta_method

    def __get__(self, obj, objtype):
        if obj is None:
            # invoked from class
            return MethodType(self.meta_method, objtype)
        else:
            # invoked from instance of class
            return MethodType(self.method, obj)


class MyClassMeta(type):
    def __new__(cls, name, bases, attrs):
        if "foo" in attrs:
            attrs["foo"] = MethodDescriptor(
                    attrs["foo"], getattr(cls, "foo")
            )
        return super().__new__(cls, name, bases, attrs)

    def foo(self):
        print("Invoking metaclass method")
        return 1


class MyClass(metaclass=MyClassMeta):
    pass


class Bar(MyClass):
    def foo(self):
        print("Invoking class method")
        return 2


# Invoking metaclass method
assert Bar.foo() == 1  # Argument missing for parameter "self" (Pyright reportGeneralTypeIssues)
# Invoking class method
assert Bar().foo() == 2
```

**VS Code extension or command-line**
CLI; `pyright 1.1.273`

**Comment by erictraut** [comment]
Your metaclass is dynamically patching the class that is passed to it. Of course, Python allows you to do such things, but this is not something a static type checker will be able to handle. If you comment out the code that patches `attrs` in the `MyClassMeta.__new__` class, you will see a runtime exception that is consistent with the error pyright emits. 

For reference, mypy also (correctly) emits the same error in this case.

My recommendation is to avoid doing hackery like this. Don't patch class attributes in your metaclass if you want your code to work with static type checking. I don't know enough about what motivated you to go down this route, so I can't offer alternative suggestions to the problem you're trying to solve.
